libgit/autogit_manager.go

```diff
@@ -0,0 +1,397 @@
+	if err != nil {
+		return err
+	}
+
+	// Construct a src repo FS.
+	srcRepoFS, _, err := GetRepoAndID(
+		ctx, gitConfig, req.srcTLF, req.srcRepo, uniqID)
+	if err != nil {
+		return err
+	}
+
+	// And a dst parent checkout FS, which better already exist.
+	dstFS, err := libfs.NewFS(
+		ctx, gitConfig, req.dstTLF, req.dstDir, uniqID,
+		keybase1.MDPriorityNormal)
+	if err != nil {
+		return err
+	}
+
+	// Take the lock for the dst repo while resetting.
+	lockFile, err := dstFS.Create(autogitLockName(req.srcRepo))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// Because we took the lock, this Close will sync/flush the
+		// whole journal.
+		closeErr := lockFile.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+	err = lockFile.Lock()
+	if err != nil {
+		return err
+	}
+
+	dstRepoFS, err := dstFS.Chroot(req.srcRepo)
+	if err != nil {
+		return err
+	}
+
+	// For now, assume the branch name refers to a ref head.
+	branch := plumbing.ReferenceName(
+		fmt.Sprintf("refs/heads/%s", req.branchName))
+	am.log.CDebugf(ctx, "Starting the reset")
+	return Reset(ctx, srcRepoFS, dstRepoFS, branch)
+}
+
+func (am *AutogitManager) clearFromQueue(req resetReq) (
+	waitCh <-chan struct{}) {
+	am.lock.Lock()
+	defer am.lock.Unlock()
+	inProgress, ok := am.resetsInProgress[req.dstPath()]
+	if ok {
+		return inProgress.doneCh
+	}
+	delete(am.resetsInQueue, req.dstPath())
+	am.resetsInProgress[req.dstPath()] = req
+	return nil
+}
+
+func (am *AutogitManager) clearFromInProgress(req resetReq) {
+	am.lock.Lock()
+	defer am.lock.Unlock()
+	delete(am.resetsInProgress, req.dstPath())
+}
+
+func (am *AutogitManager) resetWorker(wg *sync.WaitGroup) {
+	defer wg.Done()
+	for req := range am.resetQueue {
+		ctx := libkbfs.BackgroundContextWithCancellationDelayer()
+		ctx = libkbfs.CtxWithRandomIDReplayable(
+			ctx, ctxIDKey, ctxOpID, am.log)
+		for {
+			waitCh := am.clearFromQueue(req)
+			if waitCh == nil {
+				break
+			}
+			// We need to wait for any in-progress resets, since the
+			// File.Lock is taken per-instance, not per-goroutine.
+			am.log.CDebugf(ctx, "Waiting to process reset request from "+
+				"%s/%s to %s/%s",
+				req.srcTLF.GetCanonicalPath(), req.srcRepo,
+				req.dstTLF.GetCanonicalPath(), req.dstDir)
+			<-waitCh
+			am.log.CDebugf(ctx, "Done waiting")
+		}
+
+		_ = am.doReset(ctx, req)
+		am.clearFromInProgress(req)
+		close(req.doneCh)
+	}
+}
+
+func (am *AutogitManager) resetLoop(numWorkers int) {
+	var wg sync.WaitGroup
+	wg.Add(numWorkers)
+	for i := 0; i < numWorkers; i++ {
+		go am.resetWorker(&wg)
+	}
+	wg.Wait()
+	close(am.queueDoneCh)
+}
+
+func (am *AutogitManager) queueReset(ctx context.Context, req resetReq) (
+	doneCh chan struct{}, err error) {
+	dp := req.dstPath()
+	doneCh = func() chan struct{} {
+		am.lock.Lock()
+		defer am.lock.Unlock()
+		if req, ok := am.resetsInQueue[dp]; ok {
+			return req.doneCh
+		}
+		am.resetsInQueue[dp] = req
+		return nil
+	}()
+	if doneCh != nil {
+		return doneCh, nil
```

Comment:
I guess this is where "multiple outstanding requests for the same destination folder get rolled up into one" comes from. But what if `req` is different from what's in `am.resetsInQueue` already? For example it could be about "resetting" a different repo within the same `dstPath`, or "resetting" the same repo to a different ref right?

Reply:
Yeah, it's a bug that `resetReq.dstPath()` doesn't include the repo name.  Good catch, I'll fix.